### PR TITLE
add separate functions to clone model

### PR DESCRIPTION
**Pull request by sanaAyrml**
# PR Type
[Fix]

This PR separates old_model and global model cloning from set_parameters and get_parameters functions in Moon and Fenda clients.

Clickup Ticket(s): [Link](https://app.clickup.com/t/86879vn75)

# Tests Added

Fenda_client and Moon_client test has updated accordingly.

### CODE REVIEW

File: fl4health/clients/basic_client.py [hunk 1]
```diff
@@ -779,12 +781,23 @@ def get_model(self, config: Config) -> nn.Module:
         """
         raise NotImplementedError
 
+    def update_before_train(self, current_server_round: int) -> None:
+        """
+        Hook method called before training with the number of current server rounds performed.
+        For example, used by Moon and Fenda model to save global modules after aggregation.
```
Comment (lotif) [inline fl4health/clients/basic_client.py:787]: Did you mean "models" instead?

File: fl4health/clients/basic_client.py [hunk 2]
```diff
@@ -779,12 +781,23 @@ def get_model(self, config: Config) -> nn.Module:
...
+        """
+        pass
+
     def update_after_train(self, local_steps: int, loss_dict: Dict[str, float]) -> None:
         """
         Hook method called after training with the number of local_steps performed over the FL round and
         the corresponding loss dictionary. For example, used by Scaffold to update the control variates
         after a local round of training. Also used by FedProx to update the current loss based on the loss
-        returned during training.
+        returned during training. Also used by Moon and Fenda model to save trained modules weights before
```
Comment (lotif) [inline fl4health/clients/basic_client.py:799]: Same here.

File: tests/clients/test_fenda_client.py [hunk 3]
```diff
@@ -42,6 +48,91 @@ def test_getting_parameters(get_fenda_client: FendaClient) -> None:  # noqa
...
+
+    # Make sure the aggregated module is not set to train
+    assert fenda_client.aggregated_global_module.training is False
+    for param in fenda_client.aggregated_global_module.parameters():
+        assert param.requires_grad is False
+
+    # Make sure the original model is still set to train
+    assert fenda_client.model.training is True
+    for param in fenda_client.model.parameters():
+        assert param.requires_grad is True
```
Comment (lotif) [inline tests/clients/test_fenda_client.py:81]: You should unset the seed at the end or we risk having the seed set for all the tests that run after this one. See https://github.com/VectorInstitute/FL4Health/blob/aeb4d125a0c06bbddb5ce3d65c07e46a90fab444/tests/clients/test_moon_client.py#L35

File: tests/clients/test_fenda_client.py [hunk 4]
```diff
@@ -42,6 +48,91 @@ def test_getting_parameters(get_fenda_client: FendaClient) -> None:  # noqa
...
+        assert param.requires_grad is False
+    assert fenda_client.old_global_module.training is False
+    for param in fenda_client.old_global_module.parameters():
+        assert param.requires_grad is False
+
+    # Make sure the original model is still set to train
+    assert fenda_client.model.training is True
+    for param in fenda_client.model.parameters():
+        assert param.requires_grad is True
+
```
Comment (lotif) [inline tests/clients/test_fenda_client.py:134]: Same here.

File: tests/clients/test_moon_client.py [hunk 5]
```diff
@@ -35,6 +32,68 @@ def test_setting_parameters(get_client: MoonClient) -> None:  # noqa
...
+    # Make sure the globa model is not set to train
+    assert moon_client.global_model.training is False
+    for param in moon_client.global_model.parameters():
+        assert param.requires_grad is False
+
+    # Make sure the original model is still set to train
+    assert moon_client.model.training is True
+    for param in moon_client.model.parameters():
+        assert param.requires_grad is True
+
```
Comment (lotif) [inline tests/clients/test_moon_client.py:61]: And here

File: tests/clients/test_moon_client.py [hunk 6]
```diff
@@ -35,6 +32,68 @@ def test_setting_parameters(get_client: MoonClient) -> None:  # noqa
...
+    # Make sure the all old models is not set to train
+    assert moon_client.old_models_list[0].training is False
+    for param in moon_client.old_models_list[0].parameters():
+        assert param.requires_grad is False
+
+    # Make sure the original model is still set to train
+    assert moon_client.model.training is True
+    for param in moon_client.model.parameters():
+        assert param.requires_grad is True
+
```
Comment (lotif) [inline tests/clients/test_moon_client.py:95]: And here :)

**Review by lotif — CHANGES_REQUESTED** [review]
Some minor fixes.

File: fl4health/clients/basic_client.py (hunk 2 above)
Comment (sanaAyrml) [reply fl4health/clients/basic_client.py:799]: Good catch will do it.

File: fl4health/clients/basic_client.py (hunk 1 above)
Comment (sanaAyrml) [reply fl4health/clients/basic_client.py:787]: Actually I meant modules for modules as we call them global and local module in Fenda model. But I agree that model should be updated to models.

Comment (lotif) [reply fl4health/clients/basic_client.py:787]: oh ok, thanks for clarifying :)

**Review by lotif — APPROVED** [review]
Thanks for fixing the other tests too!
